transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,355 @@
+
+  private final OpenHelper openHelper;
+  private SQLiteDatabase db;
+
+  @Inject
+  SQLiteEventStore(Context applicationContext) {
+    this.openHelper = new OpenHelper(applicationContext);
+  }
+
+  private SQLiteDatabase getDb() {
+    if (db == null) {
+      db = openHelper.getWritableDatabase();
+    }
+    return db;
+  }
+
+  @Override
+  public <T> T atomically(AtomicFunction<T> function) {
+    return inTransaction(db -> function.execute());
+  }
+
+  @Override
+  public PersistedEvent persist(String backendName, EventInternal event) {
+    ContentValues values = new ContentValues();
+    values.put("backend_id", backendName);
+    values.put("transport_name", event.getTransportName());
+    values.put("priority", event.getPriority().ordinal());
+    values.put("timestamp_ms", event.getEventMillis());
+    values.put("uptime_ms", event.getUptimeMillis());
+    values.put("payload", event.getPayload());
+    values.put("num_attempts", 0);
+
+    long newRowId =
+        inTransaction(
+            db -> {
+              long newEventId = db.insert("events", null, values);
+
+              // TODO: insert all with one sql query.
+              for (Map.Entry<String, String> entry : event.getMetadata().entrySet()) {
+                ContentValues metadata = new ContentValues();
+                metadata.put("event_id", newEventId);
+                metadata.put("name", entry.getKey());
+                metadata.put("value", entry.getValue());
+                db.insert("event_metadata", null, metadata);
+              }
+              return newEventId;
+            });
+
+    return PersistedEvent.create(newRowId, backendName, event);
+  }
+
+  @Override
+  public void recordFailure(Iterable<PersistedEvent> events) {
+    StringBuilder query =
+        new StringBuilder("UPDATE events SET num_attempts = num_attempts + 1 WHERE _id in (");
+    Iterator<PersistedEvent> iterator = events.iterator();
+    while (iterator.hasNext()) {
+      query.append(iterator.next().getId());
+      if (iterator.hasNext()) {
+        query.append(',');
+      }
+    }
+    query.append(')');
+    inTransaction(
+        db -> {
+          db.compileStatement(query.toString()).execute();
+          db.compileStatement("DELETE FROM events WHERE num_attempts >= " + MAX_RETRIES).execute();
+          return null;
+        });
+  }
+
+  @Override
+  public void recordSuccess(Iterable<PersistedEvent> events) {
+    StringBuilder query = new StringBuilder("DELETE FROM events WHERE _id in (");
+    Iterator<PersistedEvent> iterator = events.iterator();
+    while (iterator.hasNext()) {
+      query.append(iterator.next().getId());
+      if (iterator.hasNext()) {
+        query.append(',');
+      }
+    }
+    query.append(')');
+    getDb().compileStatement(query.toString()).execute();
+  }
+
+  @Override
+  @Nullable
+  public Long getNextCallTime(String backendName) {
+    try (Cursor cursor =
```

Comment:
Consider using `SQLiteStatement.simpleQueryForString` for cases like this. It avoids constructing a Cursor and makes the call much less expensive.

Reply:
db.compileStatement does allow binding. See bindFoo() on SQLiteProgram

It's something like this:
```java
SQLiteStatement query = db.compileStatement("SELECT next_request_ms FROM backends WHERE name = ?");
query.bindString(1, backendName);
nextRequestMs  = query.simpleQueryForLong();
```

Reply:
cool, that's a good one to know about. I will keep my implementation in this case since I am not guaranteed that the database contains any matching rows, and this will throw an exception and I'll need to have try-catch everywhere.